packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -3,50 +3,197 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:ui';
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/painting.dart';
+import 'package:flutter/services.dart';
 
-/// A leaf node in the focus tree that can receive focus.
+import 'binding.dart';
+import 'focus_traversal.dart';
+import 'framework.dart';
+
+/// Signature of a callback used by [Focus.onKey] and [FocusScope.onKey]
+/// to receive key events.
+///
+/// The [node] is the node that received the event.
+typedef FocusOnKeyCallback = bool Function(FocusNode node, RawKeyEvent event);
+
+/// An object that can be used by a stateful widget to obtain the keyboard focus
+/// and to handle keyboard events.
+///
+/// {@template flutter.widgets.focus_manager.focus.advancedWarning}
+/// **If you are not implementing a widget which handles focused input directly
+/// (e.g. implementing your own text field or other widget primitive), or
+/// keeping [FocusNode] references so that you can control the focus from
+/// another part of the application, you typically will not need to manage a
+/// [FocusNode] directly, as it will be handled by the [Focus] and [FocusScope]
+/// widgets or by the widget primitives themselves. Please see those widgets for
+/// typical usage.**
+/// {@endtemplate}
 ///
-/// The focus tree keeps track of which widget is the user's current focus. The
-/// focused widget often listens for keyboard events.
+/// [FocusNode]s are persistent objects which form a "focus tree" that is a
+/// representation of the widgets in the hierarchy which are interested in
+/// focus. They must be managed like other persistent state, which is typically
+/// done by a [StatefulWidget]. A stateful widget that manages a focus node must
+/// call [dispose] from its [State.dispose] method.
```

Comment:
that last sentence is confusing. Does it mean that it shouldn't call it from another method? Does it mean that objects that aren't stateful widgets don't need to dispose focus nodes?

We could phrase this more as an example, e.g.

"A [FocusNode] must be disposed, by calling its [dispose] method, when it is about to be discarded. For example, a [StatefulWidget] that manages a [FocusNode] would typically create it during [initState] (or as the initializer of a final field) and call [dispose] from its own [State.dispose] method."